File: templates/go/src/domain/entity.go
package domain

import "time"

// BaseEntity represents common fields for all domain entities
type BaseEntity struct {
	ID        string    `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// Entity defines the interface for all domain entities
type Entity interface {
	GetID() string
	GetCreatedAt() time.Time
	GetUpdatedAt() time.Time
	Validate() error
}

// ValueObject defines the interface for value objects in DDD
type ValueObject interface {
	Equals(other ValueObject) bool
	Validate() error
}

File: templates/csharp/src/domain/Entity.cs
namespace DddTemplate.Domain;

using System;

/// <summary>
/// Base class for all domain entities following DDD principles
/// </summary>
public abstract class BaseEntity
{
    public string Id { get; protected set; } = Guid.NewGuid().ToString();
    public DateTime CreatedAt { get; protected set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; protected set; } = DateTime.UtcNow;

    public void MarkAsModified()
    {
        UpdatedAt = DateTime.UtcNow;
    }
}

/// <summary>
/// Interface for domain entities
/// </summary>
public interface IEntity
{
    string Id { get; }
    DateTime CreatedAt { get; }
    DateTime UpdatedAt { get; }
    void Validate();
}

/// <summary>
/// Base class for value objects in DDD
/// </summary>
public abstract class ValueObject
{
    public abstract override bool Equals(object? obj);
    public abstract override int GetHashCode();

    protected static bool EqualOperator(ValueObject left, ValueObject right)
    {
        if (left is null ^ right is null)
            return false;
        return left is null || left.Equals(right!);
    }

    protected static bool NotEqualOperator(ValueObject left, ValueObject right)
        => !EqualOperator(left, right);
}
